datafusion/expr/src/signature.rs

```diff
@@ -132,7 +138,136 @@ pub enum ArrayFunctionSignature {
     /// The first argument should be non-list or list, and the second argument should be List/LargeList.
     /// The first argument's list dimension should be one dimension less than the second argument's list dimension.
     ElementAndArray,
+    /// Specialized Signature for ArrayEqual and similar functions
     ArrayAndIndex,
+    /// Specialized Signature for ArrayEmpty and similar functions
+    Array,
+}
+
+impl ArrayFunctionSignature {
+    /// Arguments to ArrayFunctionSignature
+    /// `current_types` - The data types of the arguments
+    /// `allow_null_coercion` - Whether null type coercion is allowed
+    /// Returns the valid types for the function signature
+    pub fn get_type_signature(
+        &self,
+        current_types: &[DataType],
+        allow_null_coercion: bool,
+    ) -> Result<Vec<Vec<DataType>>> {
+        fn array_append_or_prepend_valid_types(
```

Comment:
is there any reason to create these inner functions, rather than actual functions (or maybe just inline the body of the functions into the `match` statement below)?

I find the extra level of indenting somewhat confusing

Reply:
It is only for the `match` statement below